File: plugin/src/main/kotlin/dev/betrix/superSmashMobsBrawl/models/player/PlayerDocument.kt
package dev.betrix.superSmashMobsBrawl.models.player

import dev.betrix.superSmashMobsBrawl.serializers.BannedBySerializer
import dev.betrix.superSmashMobsBrawl.serializers.InstantIsoSerializer
import dev.betrix.superSmashMobsBrawl.serializers.NullableInstantIsoSerializer
import kotlin.time.ExperimentalTime
import kotlin.time.Instant
import kotlinx.serialization.Serializable

@Serializable
data class PlayerDocument
@OptIn(ExperimentalTime::class)
constructor(
        @Serializable(with = InstantIsoSerializer::class) val firstJoinedDate: Instant,
        @Serializable(with = InstantIsoSerializer::class) val lastJoinedDate: Instant,
        val isFirstTimeOnServer: Boolean,
        val username: String,
        val uuid: String,
        var selectedKitId: String,
        val banData: BanData?
)

@Serializable(with = BannedBySerializer::class)
sealed class BannedBy {
        data object System : BannedBy()

        data class Moderator(val uuid: String) : BannedBy()
}

@Serializable
data class BanData
@OptIn(ExperimentalTime::class)
constructor(
        val isBanned: Boolean,
        val reason: String,
        @Serializable(with = NullableInstantIsoSerializer::class) val expiresAt: Instant?,
        @Serializable(with = InstantIsoSerializer::class) val bannedAt: Instant,
        val bannedBy: BannedBy
)

File: apps/backend/src/schemas/player-document.ts
import { z } from "zod";

export const playerDocumentSchema = z.object({
  isFirstTimeOnServer: z.boolean(),
  lastJoinDate: z.string(),
  headSkinBase64: z.string().nullable(),
  dailyLoginStreak: z.number().int().nonnegative().default(0),
  stats: z.record(z.string(), z.union([z.string(), z.number(), z.boolean()])),
  selectedKitId: z.string(),
  banData: z
    .array(
      z.object({
        id: z.string(),
        isBanned: z.boolean(),
        reason: z.string(),
        expiresAt: z.string().nullable(),
        bannedAt: z.string(),
        bannedBy: z.string(),
      }),
    )
    .nullable(),
});

export type PlayerDocument = z.infer<typeof playerDocumentSchema>;
